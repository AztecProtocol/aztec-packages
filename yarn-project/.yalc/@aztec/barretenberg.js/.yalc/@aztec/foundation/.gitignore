dest
node_modules
.cache
.env
*.log
*.swp
.tsbuildinfo
.pnp.cjs
.pnp.loader.mjs