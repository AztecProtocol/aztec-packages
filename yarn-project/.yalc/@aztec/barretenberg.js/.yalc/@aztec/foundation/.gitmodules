[submodule "build-system"]
	path = build-system
	url = git@github.com:AztecProtocol/build-system.git
