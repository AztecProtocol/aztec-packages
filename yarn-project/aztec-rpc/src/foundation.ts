import nodeCrypto from 'crypto';
import { Keccak } from 'sha3';

export const randomBytes = (len: number) => {
  return nodeCrypto.randomBytes(len) as Buffer;
};

export function keccak256(input: Buffer | string) {
  const inputBuf = typeof input === 'string' ? Buffer.from(input) : input;
  const hash = new Keccak(256);
  return hash.update(inputBuf).digest();
}
