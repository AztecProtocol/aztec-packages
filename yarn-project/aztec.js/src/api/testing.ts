export { AnvilTestWatcher } from '../test/anvil_test_watcher.js';
export { RollupCheatCodes } from '../test/rollup_cheat_codes.js';
export { AztecCheatCodes } from '../test/aztec_cheat_codes.js';
export { EthCheatCodes } from '@aztec/ethereum/eth-cheatcodes';
export { CheatCodes } from '../test/cheat_codes.js';
