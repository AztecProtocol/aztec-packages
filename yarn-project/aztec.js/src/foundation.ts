export function sleep(ms: number) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

export class Timer {
  private start: number;

  constructor() {
    this.start = new Date().getTime();
  }

  public ms() {
    return new Date().getTime() - this.start;
  }

  public s() {
    return (new Date().getTime() - this.start) / 1000;
  }
}

export function* backoffGenerator() {
  const v = [1, 1, 1, 2, 4, 8, 16, 32, 64];
  let i = 0;
  while (true) {
    yield v[Math.min(i++, v.length - 1)];
  }
}

export async function retry<Result>(fn: () => Promise<Result>, name = 'Operation', backoff = backoffGenerator()) {
  while (true) {
    try {
      return await fn();
    } catch (err: any) {
      const s = backoff.next().value;
      if (s === undefined) {
        throw err;
      }
      console.log(`${name} failed. Will retry in ${s}s...`);
      console.log(err);
      await sleep(s * 1000);
      continue;
    }
  }
}

// Call `fn` repeatedly until it returns true or timeout.
// Both `interval` and `timeout` are seconds.
// Will never timeout if the value is 0.
export async function retryUntil<T>(fn: () => Promise<T | undefined>, name = '', timeout = 0, interval = 1) {
  const timer = new Timer();
  while (true) {
    const result = await fn();
    if (result) {
      return result;
    }

    await sleep(interval * 1000);

    if (timeout && timer.s() > timeout) {
      throw new Error(name ? `Timeout awaiting ${name}` : 'Timeout');
    }
  }
}
