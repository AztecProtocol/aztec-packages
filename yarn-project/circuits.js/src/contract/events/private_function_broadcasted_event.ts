import { FunctionSelector, bufferFromFields } from '@aztec/foundation/abi';
import { type AztecAddress } from '@aztec/foundation/aztec-address';
import { toBigIntBE } from '@aztec/foundation/bigint-buffer';
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, type Tuple } from '@aztec/foundation/serialize';
import { type ExecutablePrivateFunctionWithMembershipProof, type PrivateFunction } from '@aztec/types/contracts';

import chunk from 'lodash.chunk';

import {
  ARTIFACT_FUNCTION_TREE_MAX_HEIGHT,
  FUNCTION_TREE_HEIGHT,
  MAX_PACKED_BYTECODE_SIZE_PER_PRIVATE_FUNCTION_IN_FIELDS,
  REGISTERER_CONTRACT_CLASS_REGISTERED_MAGIC_VALUE,
  REGISTERER_PRIVATE_FUNCTION_BROADCASTED_ADDITIONAL_FIELDS,
  REGISTERER_PRIVATE_FUNCTION_BROADCASTED_MAGIC_VALUE,
} from '../../constants.gen.js';

/** Event emitted from the ContractClassRegisterer. */
export class PrivateFunctionBroadcastedEvent {
  constructor(
    public readonly contractClassId: Fr,
    public readonly artifactMetadataHash: Fr,
    public readonly unconstrainedFunctionsArtifactTreeRoot: Fr,
    public readonly privateFunctionTreeSiblingPath: Tuple<Fr, typeof FUNCTION_TREE_HEIGHT>,
    public readonly privateFunctionTreeLeafIndex: number,
    public readonly artifactFunctionTreeSiblingPath: Tuple<Fr, typeof ARTIFACT_FUNCTION_TREE_MAX_HEIGHT>,
    public readonly artifactFunctionTreeLeafIndex: number,
    public readonly privateFunction: BroadcastedPrivateFunction,
  ) {}

  static isPrivateFunctionBroadcastedEvent(log: Buffer) {
    return toBigIntBE(log.subarray(0, 32)) == REGISTERER_PRIVATE_FUNCTION_BROADCASTED_MAGIC_VALUE;
  }

  static fromLogs(logs: { contractAddress: AztecAddress; data: Buffer }[], registererContractAddress: AztecAddress) {
    return logs
      .filter(log => PrivateFunctionBroadcastedEvent.isPrivateFunctionBroadcastedEvent(log.data))
      .filter(log => log.contractAddress.equals(registererContractAddress))
      .map(log => this.fromLogData(log.data));
  }

  static fromLogData(log: Buffer) {
    if (!this.isPrivateFunctionBroadcastedEvent(log)) {
      throw new Error(
        `Log data for PrivateFunctionBroadcastedEvent is not prefixed with magic value 0x${REGISTERER_CONTRACT_CLASS_REGISTERED_MAGIC_VALUE}`,
      );
    }

    const expectedLength =
      32 *
      (MAX_PACKED_BYTECODE_SIZE_PER_PRIVATE_FUNCTION_IN_FIELDS +
        REGISTERER_PRIVATE_FUNCTION_BROADCASTED_ADDITIONAL_FIELDS);
    if (log.length !== expectedLength) {
      throw new Error(
        `Unexpected PrivateFunctionBroadcastedEvent log length: got ${log.length} but expected ${expectedLength}`,
      );
    }

    const reader = new BufferReader(log.subarray(32));
    const event = PrivateFunctionBroadcastedEvent.fromBuffer(reader);
    if (!reader.isEmpty()) {
      throw new Error(
        `Unexpected data after parsing PrivateFunctionBroadcastedEvent: ${reader.readToEnd().toString('hex')}`,
      );
    }

    return event;
  }

  static fromBuffer(buffer: Buffer | BufferReader) {
    const reader = BufferReader.asReader(buffer);
    const contractClassId = reader.readObject(Fr);
    const artifactMetadataHash = reader.readObject(Fr);
    const unconstrainedFunctionsArtifactTreeRoot = reader.readObject(Fr);
    const privateFunctionTreeSiblingPath = reader.readArray(FUNCTION_TREE_HEIGHT, Fr);
    const privateFunctionTreeLeafIndex = reader.readObject(Fr).toNumber();
    const artifactFunctionTreeSiblingPath = reader.readArray(ARTIFACT_FUNCTION_TREE_MAX_HEIGHT, Fr);
    const artifactFunctionTreeLeafIndex = reader.readObject(Fr).toNumber();
    const privateFunction = BroadcastedPrivateFunction.fromBuffer(reader);

    return new PrivateFunctionBroadcastedEvent(
      contractClassId,
      artifactMetadataHash,
      unconstrainedFunctionsArtifactTreeRoot,
      privateFunctionTreeSiblingPath,
      privateFunctionTreeLeafIndex,
      artifactFunctionTreeSiblingPath,
      artifactFunctionTreeLeafIndex,
      privateFunction,
    );
  }

  toFunctionWithMembershipProof(): ExecutablePrivateFunctionWithMembershipProof {
    return {
      ...this.privateFunction,
      bytecode: this.privateFunction.bytecode,
      functionMetadataHash: this.privateFunction.metadataHash,
      artifactMetadataHash: this.artifactMetadataHash,
      unconstrainedFunctionsArtifactTreeRoot: this.unconstrainedFunctionsArtifactTreeRoot,
      privateFunctionTreeSiblingPath: this.privateFunctionTreeSiblingPath,
      privateFunctionTreeLeafIndex: this.privateFunctionTreeLeafIndex,
      artifactTreeSiblingPath: this.artifactFunctionTreeSiblingPath.filter(fr => !fr.isZero()),
      artifactTreeLeafIndex: this.artifactFunctionTreeLeafIndex,
    };
  }
}

export class BroadcastedPrivateFunction implements PrivateFunction {
  constructor(
    /** Selector of the function. Calculated as the hash of the method name and parameters. The specification of this is not enforced by the protocol. */
    public readonly selector: FunctionSelector,
    /** Artifact metadata hash */
    public readonly metadataHash: Fr,
    /** Hash of the verification key associated to this private function. */
    public readonly vkHash: Fr,
    /** ACIR and Brillig bytecode */
    public readonly bytecode: Buffer,
  ) {}

  static fromBuffer(buffer: Buffer | BufferReader) {
    const reader = BufferReader.asReader(buffer);
    const selector = FunctionSelector.fromField(reader.readObject(Fr));
    const metadataHash = reader.readObject(Fr);
    const vkHash = reader.readObject(Fr);
    const encodedBytecode = reader.readBytes(MAX_PACKED_BYTECODE_SIZE_PER_PRIVATE_FUNCTION_IN_FIELDS * 32);
    const bytecode = bufferFromFields(chunk(encodedBytecode, Fr.SIZE_IN_BYTES).map(Buffer.from).map(Fr.fromBuffer));
    return new BroadcastedPrivateFunction(selector, metadataHash, vkHash, bytecode);
  }
}
