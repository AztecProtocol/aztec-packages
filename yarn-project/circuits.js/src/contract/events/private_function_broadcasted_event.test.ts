import { setupCustomSnapshotSerializers } from '@aztec/foundation/testing';

import { getSamplePrivateFunctionBroadcastedEventPayload } from '../../tests/fixtures.js';
import { PrivateFunctionBroadcastedEvent } from './private_function_broadcasted_event.js';

describe('PrivateFunctionBroadcastedEvent', () => {
  beforeAll(() => setupCustomSnapshotSerializers(expect));
  it('parses an event as emitted by the ContractClassRegisterer', () => {
    const data = getSamplePrivateFunctionBroadcastedEventPayload();
    const event = PrivateFunctionBroadcastedEvent.fromLogData(data);
    expect(event).toMatchSnapshot();
  });
});
