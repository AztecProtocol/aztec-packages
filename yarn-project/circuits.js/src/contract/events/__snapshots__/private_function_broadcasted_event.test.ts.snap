// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrivateFunctionBroadcastedEvent parses an event as emitted by the ContractClassRegisterer 1`] = `
PrivateFunctionBroadcastedEvent {
  "artifactFunctionTreeLeafIndex": 5,
  "artifactFunctionTreeSiblingPath": [
    Fr<0x0aa8c40de16e14b1d7789457012932df51eab8d07d3506a2bf3c0dd0d03cab15>,
    Fr<0x03b075046b71ff5f8e0792de4b82ddc979fcb3b8c03abe12972e65c57759fb46>,
    Fr<0x1d1c34ad4b418b30c4c8b87939b97bc03d94d24d469d3cb2f419ffe6bd93f96c>,
    Fr<0x0000000000000000000000000000000000000000000000000000000000000000>,
    Fr<0x0000000000000000000000000000000000000000000000000000000000000000>,
  ],
  "artifactMetadataHash": Fr<0x16ad977a85a570bedc6785d159f137035d3bf9891042ee3af20488686aec168b>,
  "contractClassId": Fr<0x23ced3716a04d81b58822bc3e1843626aa2884888b1a2d2250e79fb7d41a365e>,
  "privateFunction": BroadcastedPrivateFunction {
    "bytecode": Buffer<0x1f8b08000000000000ffe59d057814471bc7f73609108a954285da4129dac25d04922aee4e71bd241708924012dc823b454b71eaeeeeeeded29642dddd5dbfff1b769b61d9a4d0cca4f93fdf3ecfefb9bdbdbddddf3b333b33b7b73b1bb0f64d7b6dcbbadede371f0031ceab2c0a7a96b9afea7caccfb20a3edfade2b3ac9acfb21a3ecb6afa2c3b0ab4f22c3bd167bda0cfb2ba3ecbeaf92c6be02c53a780f3daca794d0cb5484a8ab64c888613c39150426a5a4a72282939ad454a38259c9c929c91909298184d494a69999a96da32941a4e4a8c86339353133343fba62e76d1b642a59a12d24d7a76fdf79e49de05e256db293baeaba4c39fce7c43ab68be9132dfd559c7fd5e37bcef0e7a809e76d172778af1a441a87453b8b1c66d75b3f5e54d2f6d65c83f6f7a2979d35d99efa1ccf7f4e44d6fbcef03fa82737cf2c6d69c370d3496f57efad233acd627b2dd205ee394b4f54e01cde912b0f4960dd7b18633df1f310d0003c12030180c0143c130301c8c00234104a481749001a2b69318314a8130991843351e283ebabab69da0a645a6f366946deddf3ac9077f7896c94ade52a5bb161caaa1e68a663a93c65a7094a1ccd55d53e98c79b4b2ad704a624242cb44592f2523144eca484f484948c8484b0aa58722e909d1d4a4706a66524252627a467a1ab6190967863223e9a99929fbbce440747b73de49f78138dad65f2bc994651b14ceb2f56f778cc6c2602aee314a75a469bbbeae3a2a8031b6feed8ed55c58ddca5ab65bd62de008c216709cf366bcb7051ce7d3028e2f83167084c616709cc60a603c490ba833e66cd21630db500b98631b14ce31d0024e28e72da0c43d81a4051cefb8eadeee44432de0c4ffa0054c236c01739d3779de1630d7a705cc2b8316304d630b98abb102c823690175c69c4fda02e61b6a0127d90685271968012797f31650e29e4cd202e639aebab73bc5500b38e53f6801a3842de054e7cd346f0b38d5a7059c56062d6054630b38556305308da405d419f374d21670baa11670866d50788681167066396f0125ee99242de034c755f77667196a0167292d60591d78830d1d78b36d83c2b30d1c7873caf9812771cf3170e09565611b62a8b015d806850b0c14b6b9e5bcb049dc73496af951ce81a1bb5ba6b3429867731eb0c30c1db0f36d83c2f30d1cb00bcaf9012b712f206f1d861b2a6c0b6d83c20b0d14b645e5bcb049dc8b485a87f1ce81a1bb75d059212cb6cda661a87453616f659e81df514bca79dc52c6171b887b29696f60a4a10a7a996d507899810a7a7939afa025eee5e4bd8188a1c2b6c23628bcc240615b59ce0b9bc4bd92a43790e71c18ba7b033a2b8455e5bc5594567ba98156f1dc721eb794f15506e25e4dda1b48375441afb10d0aaf315041af2de715b4c4bd96bc379061a8b0adb30d0aaf3350d8d697f3c22671af27e90d4c730e0cddbd019d15c28672de2a4aabbdda40ab785e398f5bcaf80603716fd41cb73be9f63c5f9b67629249cf4dffde33c3bb40dcbc37844b3ab8377e0f56e68728f39becfd6f08df8cf75bc056b0cd367fb3bec6db41c39b35d66ddb0d9575dde9a7f15fb0f0168de9b78324fd34feb113deaa31fd766aee547aeb86ed4a1db04399dfa9cc6ff3d40d17e0fd85e022707119d40d1a6f940b5fa0316f2e2129db1acf89872fd4987e9792a49fc6d3bce18b34a6df6586eb864b943ae05265fe3265fe624fdd7039de5f01ae04579541dda0f116a2f0e51af3e66a92b2adf10c59f80a8de9770d49fa693ce913be5263fa5d6bb86eb85aa903ae51e6af55e6aff2d40dd7d9fb06afbb01dc58067583c69b2bc2d769cc9b9b0ce7cd4d4a1e5cafccdfa0ccdfe8c99b9bf1fe16702bb8cdde7f7b6a9ee83867b2d1c03989db35e6b5a54cba3defd0e6194ef3cbfb3b943ceeafcc0fb28b1f98ef4ebcbf0bdc0dee2983e37280c6bcba53e37179afe1e3f25e250fee52e6ef56e6eff1e4cd7d787f3f78003ce89337bacbe743ff3e0d120f260dd4f2395099bf4f997fc893060fe3fd23e051f05819a4c1e3faca41a69f5fa8745358eaeadb0dd4a14f182eff8f2b79fc8432ff700979ff24de3f059e06cfd8ff3c306569d320a0319f9e34d487d59defcfeaf3ccf0cbf76795fc95fc72e77b9790efcfe1fdf3e005f0e241e47ba874537890c676e4398df95e967f600fd07cfcbbd34bb641e1976cfddbdda5b130988a7b975d94c09ab66bf487b2baadd2a6e9008df9f332e9c1fab2a183f515dba0f02b060ed657cbf9c12a71bf6ae860d5dd13d8e5b8ea6e5d77d9e533bfe3aca25e49719326ef44d94f8c67dbbacb5a434b7f63a0dbb111816363038e8543264841ab601d38e90e40e3cf18238306252727256646232d4da641ef32faf9152add14ee47e2196b95ef32e54e6a65bedb79f39aed24883b8e897c10f408e86ef5766b6ca9f6d87a2b07371df6f8a483ee82f39aa102ae3bbff668ccafbd86f26bef41945b1dbd605dee4dacb2c9fbf274deaf2949ccb6c6984f2189394663cca792c4acb1e10e372ba39843a59bc2cd35a61f4b473264717886493c13483c13493c93483c93493c5b9078b624f14c21f14c25f13c8dc4f37412cf33483ccf24f13c8bc4f36c12cf56249ead493cdb9078b625f16c47e2d99ec4b30389674712cf4e249e9d493cbb90787625f1ec46e2d99dc4b30789674f12cf5e249ebd493cfb9078f625f13c87c4b31f89677f12cf01249e03493c0791780e26f11c42e23994c4731889e77012cf11249e23493c23249e69249ee9249e19249e5112cf4c12cf51249ea3493cb3483cc790788e25f11c47e2399ec4339bc43387c4730289e74412cf5c12cf3c12cf7c12cf49249e93493ca790784e25f19c46e2399dc4730689e74c12cf59249eb3493ce79078169078ce25f19c47e2399fc4730189e74212cf45249e8b493c9790782e25f15c46e2b99cc4730589e74a12cf55249ee79278ae26f15c43e2b996c4731d89e77a12cf0d249ee791786e24f13c9fc4731389e76612cf2d249e5b493cb791786e27f1dc41e2b993c4f30212cf0b493c2f22f1bc98c4f31212cf4b493c2f23f1bc9cc4f30a12cf2b493caf22f1bc9ac4f31a12cf6b493caf23f1bc9ec4f30612cf1b493c6f22f1bc99c4f31612cf5b493c6f23f1bc9dc4f30e12cf3b493cef22f1bc9bc4f31e12cf7b493cef23f1bc9fc4f30112cf07493c1f22f17c98c4f31112cf47493c1f23f17c9cc4f30912cf27493c9f22f17c9ac4f31912cf67493c9f23f17c9ec4f30512cf17493c5f22f1dc45e2f93289e72b249eaf9278ee26f17c8dc4730f89e75e12cfd7493cdf20f17c93c4f32d12cfb7493cdf21f17c97c4f33d12cff70d79da1ecff2f42cee0fca28e650e9a6f0871ad34f7d8e7da89493c9f2f891c5e1f93189e727249e9f92787e46e2f93989e717249e5f92787e45e2f93589e737249edf92787e47e2f93d89e70f249e3f9278fe44e2f93389e72f249ebf9278fe46e2f93b89e71f249e7f9278fe45e2291b64f00c9078da249e31249eb1249e71249e15483c2b92785622f18c27f1ac4ce2791889671512cfaa249ed5483cab9378d620f13c9cc4b32689e711249eb50c7996e7eb026aff1fc67ce4ff61cc47911c8347074a9f7ea991b4f450383962d2f318923214ab31e63a243137d718f3b124c7cd71659437a1d24de1e303faf2b99fcd11f3091a63de4d724ddb8924c74d90c4b32e89673d12cf93483ceb93789e4ce2d980c4b32189672312cfc6249e4d483c9b92789e42e2792a89673312cfe6249e2112cf3089670289672289671289673289670b12cf96249e29249ea9249ea791789e4ee2790689e799249e6791789e4de2d98ac4b33589671b12cfb6249eed483cdb93787620f1ec48e2d989c4b33389671712cfae249edd483cbb9378f620f1ec49e2d98bc4b33789671f12cfbe249ee79078f623f1ec4fe23980c4732089e72012cfc1249e43483c8792780e23f11c4ee23982c47324896784c4338dc4339dc43383c4334ae29949e2398ac4733489671689e71812cfb1249ee3483cc79378669378e690784e20f19c48e2994be29947e2994fe23989c4733289e71412cfa9249ed3483ca79378ce20f19c49e2398bc4733689e71c12cf0212cfb9249ef3483ce793782e20f15c48e2b988c4733189e71212cfa5249ecb483c979378ae20f15c49e2b98ac4f35c12cfd5249e6b483cd79278ae23f15c4fe2b981c4f33c12cf8d249ee793786e22f1dc4ce2b985c4732b89e73612cfed249e3b483c7792785e40e2792189e745249e1793785e42e2792989e765249e9793785e41e2792589e755249e5793785e43e2792d89e775249ed79378de40e2792389e74d249e379378de42e2792b89e76d249eb79378de41e2792789e75d249e779378de43e2792f89e77d249ef793783e40e2f92089e743249e0f93783e42e2f92889e763249e8f93783e41e2f9a4214fdbe359dae73036d018f353651473a87453f8e980bef4db6b7394c767488e9b67493c9f23f17c9ec4f30512cf17493c5f22f1dc45e2f93289e72b249eaf9278ee26f17c8dc4730f89e75e12cfd7493cdf20f17c93c4f32d12cfb7493cdf21f17c97c4f33d12cff7493c3f20f1fc90c4f32312cf8f493c3f21f1fc94c4f33312cfcf493cbf20f1fc92c4f32b12cfaf493cbf21f1fc96c4f33b12cfef493c7f20f1fc91c4f32712cf9f493c7f21f1fc95c4f33712cfdf493cff20f1fc93c4f32f124f8be4fa900089a74de21943e2194be21947e25981c4b32289672512cf7812cfca249e8791785621f1ac4ae2598dc4b33a89670d12cfc3493c6b92781e41e2598bc4b33689e791249e4791781e4de2790c89671d12cf63493c8f23f13c9ec4f30412cf13493c83249e75493ceb91789e44e2599fc4f36412cf06249e0d493c1b19f2b43d9ea5bd0f3a4e63cc8d4962aea031e626243157d418735392982b698cf9149298e335c67c2a49cc9535c6dc8c24e6c334c6dc9c24e62a1a630e91c45c5563cc619298ab698c398124e6ea1a634e2489b986c6989348623e5c63ccc92431d7d418730b92988fd018734b92986b698c398524e6da1a634e2589f9488d319f4612f3511a633e9d24e6a335c67c0649ccc7688cf94c9298eb688cf92c92988fd518f3d924311fa731e65624311faf31e6d624319fa031e63624319fa831e6b624310735c6dc8e24e6ba1a636e4f12733d8d31772089f9248d31772489b9bec6983b91c47cb2c6983b6b8c39d6da774d40d009b809680a4e01a78266a0b9ec07844182a4054802c9a005680952402a380d9c0ece006782b3c0d94eecad411bd016b403ed4107d01174029d4117d0157403dd410fd013f402bd411fd0179c03fa81fe600018080681c16008180a8681e16004180922200da4830c1005996014180db2c01830168c03e34136c80113c044900bf2403e9804268329602a9806a6831960269805668339a000cc05f3c07cb0002c048bc062b0042c05cbc072b002ac04abc0b960355803d68275603dd800ce031bc1f96013d80cb680ad601bd80e76809de0027021b8085c0c2e019782cbc0e5e00a7025b80a5c0dae01d782ebc0f5e0067023b809dc0c6e01b782dbc0ede00e7027b80bdc0dee01f782fbc0fde001f02078083c0c1e018f82c7c0e3e009f024780a3c0d9e01cf82e7c0f3e005f0227809ec022f8357c0ab6037780dec017bc1ebe00df026780bbc0dde01ef82f7c0fbe003f021f8087c0c3e019f82cfc0e7e00bf025f80a7c0dbe01df82efc0f7e007f023f809fc0c7e01bf82dfc0efe00ff027f80bc8411600368801b1200e5400154125100f2a83c3401550155403d5410d7038a8098e00b5406d7024380a1c0d8e0175c0b1e038703c38019c0882a02ea8074e02f5c1c9a00168081a81c6a009680a4e01a78266a0390881304800892009248316a0254801a9e034703a38039c09ce0267cbbdf0a0356803da8276a03de8003a824ea033e802ba826ea03be8017a825ea037e803fa8273403fd01f0c0003c12030180c0143c130301c8c00234104a481749001a220138c02a341161803c68271603cc806396002980872411ec80793c06430054c05d3c0743003cc04b3c06c30071480b9601e980f1680856011580c9680a56019580e56809560153817ac066bc05ab00eac071bc0796023381f6c029bc116b0156c03dbc10eb013c8b3ede5b9f1f24c7679deb93c4b5c9ed32dcfc096e74bcbb39be5b9c8f2cc61799eaf3c2b579e432bcf7895e7a7cab349e5b99ff24c4d795ea53c0b529eb328cf3094e703cab3f7e4b976f2cc38791e9b3ceb4c9e2326cfe8ba1fc8b3a5e4b94df24c2479de903ccb479e9323cfa091e7bbc8b353e4b924f29c0e7906863c13429eb720cf3290e704c818fc32bebd8c1d2fe3b2cb98e7329eb88cd52de360cb18d3327eb38c8d2ce30ecb98be325eae8c45fb36903154657c5219fb53c6d594312b653c48196b51c63194310265fc3d19db4ec68d9331d964bc33194b4cc6e99231b0647c2919bb49c645923187643c1f192b47c6a191315e64fc14199b44c6fd90313564bc0a190b423a143286818c0f20f7decb7ded72cfb8dc8f2df73acb7dc4728faedcff2af796ca7d9b724fa4dc6f28f7f2c97d72720f9adcdf25f74ec97d4972cf8fdc4f23f7aac87d20728f85dcbf20f706c875f7724dbb5c2f2ed7624bfb2ed710cbf5b972edab5c572ad76ccaf59072ada15cc727d7b5c9755e72dd935c0724d7c5c8752272dd845c4720ffabcbffccf2bfabfc0f29ffcbc9ff54f2bf8dfc8f21e7f5e53cb79cf795f3a0725e50ce93159e3702725e417e67cbef4ef91d26bf4ba49f2efd56e9c749bfc6de57555ad22f91a989553439c968c5389fcb357d728d9b5cf325d740c93541728d8c5c3322d750c83505f21fbbfce72cffc1ca7f92f21f9dfc6725ffe1c87f1a728e5fce79cb396039272ae708e59c999c4392732a728e417e7307415d500f489f5dfab0d2a793e7d334048d4063ebc0a9b2325f4b993fce798de4e747c74fc80fe6e704231919c12959f9a3833993a3b999e372a45b51d87f70a77ace6be7ecacfcacc8b8ace991fcac9cece0e848dee860464e342f989d931f1c1fc94f97ce4c61cfc29d42de6f46736577b9d1bcbc60d6beefe58f8e06d373b2f37323e9f9c18ce8847139d3a2d2d929ec8d1caaf5aa43fc4ec0c90dcb494d77aae6bcf6cd8fa48fc53773601695aab8b0cf2453b3835a7b89f349eb435abbb3b2f631ee16727323d38259d919d1a9c19c49f9c19ccc605acea4ec8c3cf58b7d0e31fcc2922c935b82e29d57f757887b44b472de874a3785e395edeade7e4a283929deda7fd2ec9f186f15d506a6d247b6196366fb8557a9cad4b6a068fb6a2cee7eab3aef034a5abadf911aaf86f21d751b96e3ebfdaefbb9c9f852424949f1d6feb5b8def40b175e2d26534cc181e9e77e165be09f9e32c515ecefe5a6a77b944b0bd1ce59a786f2797d653bdef40e9441dacac56d01cffe2c6bff74f6c65cd9327bac043cfb737dbce9e3ce5771d257a651d1fcb64e9bd3393b2f3f929d1e554350375bc327443509fcd6f1aee75dbfaacfe7aea67422dc6aac6241d1e73aab0177bf959cedc7594559e9eeb370b9336f7bd657e703cef7d565272971c914af7ccfdd4f8cb2ccdd6775e77d05e5b3f8827fde56accfb66af8ac5fc1b3ad8a3edb5297b9714b67af8e33afe64facb23d5df9d332a5287fdc3452f3479d6295cfd5f51b39ef6b5afe93c4e076cdb89bfca49665d5e41b6a920d3759fe4dbe37164b59e6aee7fd8eac5752b7c064d7c87cd35e94cf15cc6c3fc14db73825ed627d6272f75fd53ab0790b14f36a59fe4db1bbafca96d1321c2e2936d5df2d1b353cebab69f05fe6839a3e6a3ec874b05dbe38cf67c575f9ea28eb15d70d97f54e76e6a5ebd2d1d996a5a4954cacfd84fa4acc85f128df63ed271cebcc4b1b2bf9d7d42ada5671e5ceafbb5add3ab06e88f36ccbafeb1be7b32df7a4944c1370d6a175eea849e3a3d9f979de0eabba03cb6743eacedde5c53528deefc42aebab134387aab9f3bea40e557f679ebc439556561daa8a66b6dfc2dd7e2533dbf7ed50a9b1b8fb758f8976054569d9ae607f27779d0eca3a1d8a5947adfc3b2aeba8c75f27659d4e9e754aeabcf99d5768e7f94cad6cda7b3e538fa50e9ecfd40aa9a3e7b30ad681be927eee196bb7d2ea5cb07f2cea675d94cf5cff2394effcdd2859fbae0a50f75fb168f7079c4731547e120fb6f3e6eebfb265f478f9bbf356c9e3e34d1fbff328e99171e37ae5664d8ee4473b4cca4e97bf05d410d4cdda3e2116d7cca8c5c2bb9eb798faf591633defe33cefddedabfb29eefbde65aa8f3b5fdd677945cf3adeecf62ef3f68d2bfaeccfef100e14f3eaeec7fa87fdf83997749af150f763fa94760d8fabbaafea3efbf6a6ab5f1faaa4df2c157df663f8b769b25f55ea4edeaad4afbca8cedee3d0f2c4af4e6a4c721cb755b6eb5dcfebe3f75b2ac613875fb988b1ac62fbbdea36fcba9cee7cbccf7e747631fc9a2f6ffc251dbf6ad372a8f9f1f76f28ebd0f2434d23ef71ae1ed3debf80bcdd589918baec6d9cf72575d9d52e815f5d11f07ca6a693b77e3354d6fe3e5fe19635371dbc3e71cae7eafa5dfe211d02d681c75849f5a0d7475ddffbfb5426bfdfb1153c9f9594aeffc5392bbfe3c11b7705ebd0f2a1aff35a5c3ef8b563eab924ef7ed463b19292663255b68ca45942bcb24f03db0f55b40e9cd40b590ef37ce6a6955fbd1028e6bded792d69dd4009dbadeaf399faf3c3bb2cdef35a4bd99e89d303b5cd6cdf378fd40b8c6a7be254d3b9952607777beeb1e9d72fb23d9fa9e75455bf807ebfb0d725c6675feea49615f7f3ff018d8be2fd060d0200>,
    "metadataHash": Fr<0x1eef7f2eaafa09e24b0475a918bd7388c4dec305145849f5f84d1b822202b944>,
    "selector": Selector<0xd9e0e7f1>,
    "vkHash": Fr<0x0000000000000000000000000000000000000000000000000000000000000000>,
  },
  "privateFunctionTreeLeafIndex": 5,
  "privateFunctionTreeSiblingPath": [
    Fr<0x00a5215d923c08e46ef8298acff37a856da61be92fe1536dd7b4e10cfec9590c>,
    Fr<0x233d10e0c959338e062b1c76fb81e17fcf61643da568eac372bb7feac8bfa357>,
    Fr<0x17720805123feb41cde00457a83a9919a315c55d6117267f05edefe119df9a7a>,
    Fr<0x1a0ca5eecb1430479902264f04e557160a4666fb42bb8b283c6397e3d17ac937>,
    Fr<0x2a6595890719fef7967c209488728aa5342438ba52058a3c770202f55acf6854>,
  ],
  "unconstrainedFunctionsArtifactTreeRoot": Fr<0x049fe6a55fa21b718224c0f6e69887dd82e302b6b658c9db0a389672582a343d>,
}
`;
